Pass on the HTML chapter (#7)
pavpanchekha/emberfox: book/html.md

next: layout
...

So far, your web browser sees web pages as a stream of open tags,
close tags, and text. But HTML is actually a tree, and though the tree
structure hasn't been important yet, you'll need it to draw

> backgrounds, add margins, and let alone implement CSS. So this chapter

Reviewer: ```suggestion
backgrounds, add margins, and implement CSS. So this chapter
```

    raise Exception("Reached last token before end of document"))
```

Now it's easy to see the result of parsing an HTML document:
Python prints a list of `ElementNode` objects, meaning that there were
open HTML elements still around when it reached the last token:

```
[<__main__.ElementNode object at 0x101399c70>,  ...]
```

Ok, that's not too helpful. Python needs a method called `__repr__` to
be defined to print things a little more reasonably:

``` {.python}
print_tree(parse(lex(" ... ")))
class ElementNode:
    # ...
    def __repr__(self):
        return "<" + self.tag + ">"
```

Make sure to try this for several documents. Try documents without
text, or without tags; documents without close tags; documents with
extra whitespace before or after the root element; and so on. Most
likely, you'll find incorrect results or crashes.
This produces a more reasonable result:

``` 
[<!DOCTYPE html>,
 <html lang="en-US" xml:lang="en-US">,
 <head>,
 <meta charset="utf-8" />,
 <meta name="generator" content="pandoc" />,
 <meta name="viewport" content=... />,
 <link rel="prev" href="text" />,
 <link rel="next" href="layout" />,
 <link rel="stylesheet" href="../book.css" />]
```

> When you do, the best way to get more insight is to print the state of

Reviewer: Why remove this advice?

> It doesn't help; why? Because the problem tags aren't just `<meta>`:

Reviewer: What do you mean by "it" here?

> whitespace,[^because-unquoted] we can split the tag contents on

Reviewer: The contents of an attribute value can contain spaces. Example: the sandbox attribute.
https://html.spec.whatwg.org/#sandboxing

Reviewer: You could say instead that for simplicity we are going to ignore attribute values that allow spaces, because those require handling quoted text.

Author: Yeah, it's for simplicity. In an earlier draft I pushed a 7-state parsing automaton to handle quoted values, but it was just too long. I'll clarify that.

if "=" in attrpair:
    if len(value) > 2 and value[0] in ["'", "\""]:
        value = value[1:-1]
    # ...
```

This conditional checks the first character of the value to determine
if it's quoted, and if so strips off the first and last character,
leaving the contents of the quotes.

::: {.further}
This is [not the right way][case-hard] to do case
insensitive comparisons; the Unicode case folding algorithm should be
used if you want to handle languages other than English. But in HTML
specifically, tag names only use the ASCII characters where this test
is sufficient.
:::
    
[case-hard]: https://www.b-list.org/weblog/2018/nov/26/case/

Doctype declarations
====================

Self-closing elements are now handled correctly and the list of open
elements when parsing this page is now shorter:

```
[<!DOCTYPE>]
```

[html5-doctype]: https://html.spec.whatwg.org/multipage/syntax.html#the-doctype

This special tag is called a [doctype][html5-doctype], and it's not
really an element at all. It's not supposed to have close tag, but we
can't mark it a self-closing tag—it's always the very first thing in
the document, so there wouldn't be an open element to append it to.

> It's best to throw it away:

Reviewer: ```suggestion
It's best to throw it away for now (a real browser needs to parse it to handle various legacy or obscure parsing modes[1]):
```
[1] https://en.wikipedia.org/wiki/Quirks_mode

Reviewer: Oh I see you refer to quirks mode below. Maybe just foreshadow in a parenthetical?

Author: Worth footnoting.

Author: You know, looking over the HTML5 spec again, maybe the right move here is to only drop whitespace text. HTML5 mandates that in standards mode (https://html.spec.whatwg.org/multipage/parsing.html#the-before-html-insertion-mode). And then implicit tags later in the chapter handle the case of text being inserted here.

Author: Oh, sorry, I got the context wrong here. I'll add a footnote.

browsers insert them automatically. To do so, they compare the current
token's tag to the list of currently open elements; that reveals
whether any additional elements need to be created:

``` {.python indent=4}
for tok in tokens:
    implicit_tags(tok, currently_open)
    # ...
```

The `implicit_tags` function examines the incoming token and
determines whether any implicit tags need to be added. Importantly,
more than one implicit tag may need to be added, so the function uses
a loop:

``` {.python}
def implicit_tags(tok, currently_open):
    tag = tok.tag if isinstance(tok, Tag) else None
    while True:
        open_tags = [node.tag for node in currently_open]
        # ...
```

That loop needs to handle each possible implicit tag. The implicit
`<html>` tag is easiest, since it must be the root element:

``` {.python indent=8}
if open_tags == [] and tag != "html":
    currently_open.append(ElementNode("html"))
```

With the `<head>` and `<body>` elements, you need to look at the tag
being handled. Some tags go in the `<head>` element:

``` {.python}
HEAD_TAGS = [
    "base", "basefont", "bgsound", "noscript",

>     "link", "meta", "title", "style", "script",

Reviewer: script can go in body; I guess it's only when these tags are encountered outside of body/head that this behavior happens?

Author: Yeah, but if you encounter script before either head or body is opened, HTML5 parsing will create a head node. HTML5 more or less enforces the requirement that every DOM tree starts with HTML, HEAD, and BODY, and that HEAD only contains one of those elements (except under noscript), but as you say BODY can also contain script and noscript.

```

If you see an element that's not supposed to go in the `<head>`, you
need to implicitly close the `<head>` section:

``` {.python indent=8}
elif open_tags == ["html", "head"] and tag not in ["/head"] + HEAD_TAGS:
    node = currently_open.pop()
    currently_open[-1].children.append(node)
    currently_open.append(ElementNode("body"))
```

In all other cases, implicit tags aren't needed, and we can exit the
loop:

``` {.python indent=8}
else:
    break
```

We haven't yet handled the implicit close tags `</body>` and
`</html>`. Usually, leaving them out will mean the `parse` function
reaches the end of its loop without closing all open tags, and would
thus return nothing. Let's make it return *something* instead:

``` {.python}
def parse(tokens):
    # ...
    while currently_open:
        node = currently_open.pop()
        if not currently_open: return node
        currently_open[-1].children.append(node)
```

These rules for malformed HTML may seem arbitrary, and they are: they
evolved over years of trying to guess what people "meant" when they
wrote that HTML, and are now codified in the [HTML parsing
standard](html5-parsing).

::: {.further}

> HTML parsers also have an [algorithm][adoption] to handle misnested

Reviewer: ```suggestion
HTML parsers also have an [algorithm][adoption] to handle mis-nested
```

elements, plus a [list of active formatting
elements][html5-formatting-list] to handle formatting like
`<b>b<i>bi</b>i</i>`.
:::

[html5-formatting-list]: https://html.spec.whatwg.org/multipage/parsing.html#the-list-of-active-formatting-elements
[adoption]: https://html.spec.whatwg.org/multipage/parsing.html#adoption-agency-algorithm

Summary
=======

This chapter taught our browser that HTML is a tree, not just a flat

> list of tokens:

Reviewer: ```suggestion
list of tokens. We added:
```

- A parser now transforms HTML tokens to a tree

> - Layout operates recursively on the tree

Reviewer: ```suggestion
- Layout operating recursively on the tree
```

> - The lexer recognizes and handles attributes on elements

Reviewer: ```suggestion
- Lexer code to recognize and handle attributes on elements
```

(did you introduce what a "lexer" is?)

Author: Good point! I call the function `lex` but don't think I ever called it a lexer.

> - Some malformed HTML documents are automatically fixed

Reviewer: ```suggestion
- Fixes for some malformed HTML documents
```

> The tree structure of HTML is essential to lay out visually complex

Reviewer: ```suggestion
The tree structure of HTML is essential to display visually complex
```

::: {.signup}
:::

Exercises
=========

*Entities:* Implement support for `&amp;`, `&lt;`, `&gt;`, and
`&quot;`, which expand to ampersand, less than, greater than, and the
quotation mark in HTML. These "entities" can go either in ordinary
text or inside tags. Should you handle entities in the lexer or the
parser?

*Comments:* Update the HTML lexer to support comments. Comments in
HTML begin with `<!--` and end with `-->`. However, comments aren't
the same as tags: they can contain any text, including left and right
angle brackets. The lexer should skip comments, not generating any
token at all. Test: is `<!-->` a comment, or does it just start one?

*Paragraphs:* Since it's not clear what it would mean for one
paragraph to contain another, the most common reason for this to
happen in a web page is that someone forgot a close tag. Change the
parser so that a document like `<p>hello<p>world</p>` results in two
sibling paragraphs instead of one paragraph inside another.

*Scripts:* JavaScript code embedded in a `<script>` tag uses the left
angle bracket to mean less-than. Modify your lexer so that the
contents of `<script>` tags are treated specially: no tags are allowed

> inside `<script>`, except the `</script>` close tag.

Reviewer: Fun fact: there are some weird restrictions on presence of the string "</script>" in an inline script, for reasons of preventing the HTML parser from needing to understand script syntaxes.

Author: Technically it's just `</script` followed by a space, tab, \v, \r, slash, or greater than sign. Yeah, I've run into this once or twice and it's a fun edge case. I talk about it in a video.

https://html.spec.whatwg.org/multipage/parsing.html#script-data-end-tag-name-state

        currently_open[-1].children.append(node)
```

These rules for malformed HTML may seem arbitrary, and they are: they
evolved over years of trying to guess what people "meant" when they
wrote that HTML, and are now codified in the [HTML parsing
standard](html5-parsing).

::: {.further}
HTML parsers also have an [algorithm][adoption] to handle mis-nested
elements, plus a [list of active formatting
elements][html5-formatting-list] to handle formatting like
`<b>b<i>bi</b>i</i>`.
:::

[html5-formatting-list]: https://html.spec.whatwg.org/multipage/parsing.html#the-list-of-active-formatting-elements
[adoption]: https://html.spec.whatwg.org/multipage/parsing.html#adoption-agency-algorithm

Summary
=======

This chapter taught our browser that HTML is a tree, not just a flat
list of tokens. We added:

- A parser to transform HTML tokens to a tree
- Layout operating recursively on the tree
- Code to recognize and handle attributes on elements
- Automatic fixes for some malformed HTML documents

The tree structure of HTML is essential to display visually complex
web pages, as we will see in the [next chapter](layout.md).

::: {.signup}
:::

Exercises
=========

> *Entities:* Implement support for `&amp;`, `&lt;`, `&gt;`, and

Reviewer: Rework this paragraph to refer back to the exercise I added in the other chapter, and
make this exercise about adding the other entities not in the other exercise (if the
student had already done the other exercise).

Author: Great point. I ended up eliminating this exercise and replacing it with something I think is more interesting: turning your lexer into a syntax highlighter which you'd run for `view-source:` URLs.

=======

This chapter taught our browser that HTML is a tree, not just a flat
list of tokens. We added:

> - A parser now transforms HTML tokens to a tree

Reviewer: ```suggestion
- A parser transforming HTML tokens to a tree
```
